src/ispc/kernels/lanczos3.ispc

```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
     return sin(x) / x;
 }
 
-static inline float lanczos3_filter(float t)
+static inline float lanczos3_filter(float t, uniform float filter_scale)
 {
     t = abs(t);
 
-    if (t < 3.0)
-        return clean(sinc(t) * sinc(t / 3.0));
+    if (t < filter_scale)
+        return clean(sinc(t) * sinc(t / filter_scale));
     else
         return 0.0f;
 }
 
-struct Image {
-    uniform const uint8* data;
-    uniform int<2> size;
+struct WeightVariables {
+    float src_center;
+    float src_start;
+    float src_end;
 };
 
-static inline int<2> uv_to_pixel_id(int<2> image_size, float<2> uv) {
-    float<2> pixel_coord = uv * image_size;
-    int<2> i = {floor(pixel_coord.x + 0.5), floor(pixel_coord.y + 0.5)};
-    i.x = clamp(i.x, 0, image_size.x - 1);
-    i.y = clamp(i.y, 0, image_size.y - 1);
-    return i;
+export void calculate_weight_variables(uniform float filter_scale, uniform uint32 src, uniform uint32 target, uniform WeightVariables out_variables[]) {
+    uniform float ratio = src / target;
+
+    uniform float filter_radius = ceil(ratio * filter_scale);
+
+    foreach(pixel = 0 ... target) {
+        WeightVariables vars;
+
+        vars.src_center = (pixel + 0.5) * ratio - 0.5;
+
+        float start = ceil(vars.src_center - filter_radius);
+        vars.src_start = min(max(start, 0), src - 1);
+
+        float end = floor(vars.src_center + filter_radius);
+        end = min(max(end, 0), src - 1);
+        vars.src_end = max(end, (float)vars.src_start);
+
+        out_variables[pixel] = vars;
+    }
 }
 
-static inline float frac(float f) {
-    float absf = abs(f);
-    return absf - floor(absf);
+export void calculate_weights(uniform float image_scale, uniform float filter_scale, uniform const WeightVariables * uniform vars, uniform float * uniform weights) {
+
+    uniform float start = vars->src_start;
+    uniform float end = vars->src_end;
+    uniform float center = vars->src_center;
+
+    // Calculate the Lanczos3 weight of each pixel
+    foreach(i = 0 ... end - start + 1) {
+        weights[i] = lanczos3_filter((start + (float)i - center) / image_scale, filter_scale);
+    }
+
+    // The needs to be done in a normal loop because we cannot apply vectorization to it.
+    float sum = 0.0;
+    for(int i = 0; i < end - start + 1; i++) {
+        sum += weights[i];
+    }
+
+    // Normalize the weights, such that their sum is 1.0f
+    foreach(i = 0 ... end - start + 1) {
+        weights[i] /= sum;
+    }
 }
 
-static inline float byte_to_float(uint b) {
-    //return floatbits(0x3f800000 | (b << (23 - 8))) - 1.0;
-    return (float)b;
+struct WeightCollection {
+    uniform const uint32* starts;
+    uniform const uint32* weight_counts;
+    uniform const float* const* values;
+};
+
+struct Cache {
+    uniform WeightCollection vertical_weights;
+    uniform WeightCollection horizontal_weights;
```

Comment:
I don't think this counts as a "cache" (since we don't insert nor evict anything) but only serves the precomputed weights.